lib/options_validator.js

```diff
@@ -138,4 +175,78 @@ function invalidate(validationLevel, message, logger) {
   }
 }
 
-module.exports = { createValidationFunction };
+/**
+ * This is a class to build an operation by checking its arity and validating its options.
+ */
+class OperationBuilder {
+  constructor() {
+    this.requiredArity = -1;
+    this.validationSchema = {};
+  }
+}
+
+/**
+ * Build an OperationBuilder given an operation's required arity.
+ *
+ * @method
+ * @param {number} requiredArity The required arity of the operation to build.
+ * @return {OperationBuilder} An operation builder that will validate options.
+ */
+function arity(requiredArity) {
+  const operationBuilder = new OperationBuilder(requiredArity);
+  return operationBuilder.arity(requiredArity);
+}
+
+OperationBuilder.prototype.arity = function(requiredArity) {
+  this.requiredArity = requiredArity;
+  return this;
+};
+
+/**
+ * Set an OperationBuilder's validationSchema.
+ *
+ * @method
+ * @param {object} validationSchema The schema against which to validate options.
+ * @return {OperationBuilder} An operation builder that will validate options.
+ */
+OperationBuilder.prototype.options = function(validationSchema) {
+  this.validationSchema = validationSchema;
+  return this;
+};
+
+/**
+ * Build an operation by checking its arity and validating its options.
+ *
+ * @method
+ * @param {function} operationToBuild The operation to validate.
+ * @return {function} The operation with validated options.
+ */
+OperationBuilder.prototype.build = function(operationToBuild, overrideOptions, validationOptions) {
+  const operationBuilder = this;
+
+  function buildOperation() {
```

Comment:
slightly more than a nit: in backtraces this is going to be the thing people see first, so I think it behooves us to be as clear as possible. I think this should be something more generic like `operation` or at very least `builtOperation`. You technically are building it _here_, but what you are returning is the compiled wrapped operation. 

/cc @daprahamian